libraries/chain/steem_evaluator.cpp

```diff
@@ -445,35 +482,57 @@ namespace golos { namespace chain {
-            database &_db;
+            const comment_object& _c;
+            database& _db;
+            const account_object& _a;
 
-            void operator()(const comment_payout_beneficiaries &cpb) const {
+            void operator()(const comment_payout_beneficiaries& cpb) const {
                 if (_db.is_producing()) {
                     GOLOS_CHECK_LOGIC(cpb.beneficiaries.size() <= STEEMIT_MAX_COMMENT_BENEFICIARIES,
-                            logic_exception::cannot_specify_more_beneficiaries,
-                            "Cannot specify more than ${m} beneficiaries.", ("m", STEEMIT_MAX_COMMENT_BENEFICIARIES));
+                        logic_exception::cannot_specify_more_beneficiaries,
+                        "Cannot specify more than ${m} beneficiaries.", ("m", STEEMIT_MAX_COMMENT_BENEFICIARIES));
                 }
 
-                GOLOS_CHECK_LOGIC(_c.beneficiaries.size() == 0,
+                uint16_t total_weight = 0;
+
+                if (_c.beneficiaries.size() == 1
+                        && _c.beneficiaries.front().account == _a.referrer_account) {
+                    total_weight += _c.beneficiaries[0].weight;
+                    auto& referrer = _a.referrer_account;
+                    const auto& referrer_find = std::find_if(cpb.beneficiaries.begin(), cpb.beneficiaries.end(),
+                            [&referrer](const beneficiary_route_type& benef) {
+                        return benef.account == referrer;
+                    });
+                    GOLOS_CHECK_LOGIC(referrer_find == cpb.beneficiaries.end(),
+                        logic_exception::beneficiaries_should_be_unique,
+                        "Comment already has '${referrer}' as a referrer-beneficiary.", ("referrer",referrer));
+                } else {
+                    GOLOS_CHECK_LOGIC(_c.beneficiaries.size() == 0,
                         logic_exception::comment_already_has_beneficiaries,
                         "Comment already has beneficiaries specified.");
+                }
+
                 GOLOS_CHECK_LOGIC(_c.abs_rshares == 0,
-                        logic_exception::comment_must_not_have_been_voted,
-                        "Comment must not have been voted on before specifying beneficiaries.");
+                    logic_exception::comment_must_not_have_been_voted,
+                    "Comment must not have been voted on before specifying beneficiaries.");
 
-                _db.modify(_c, [&](comment_object &c) {
-                    for (auto &b : cpb.beneficiaries) {
+                _db.modify(_c, [&](comment_object& c) {
+                    for (auto& b : cpb.beneficiaries) {
                         _db.get_account(b.account);   // check beneficiary exists
                         c.beneficiaries.push_back(b);
+                        total_weight += b.weight;
                     }
                 });
+
+                GOLOS_CHECK_LOGIC(total_weight <= STEEMIT_100_PERCENT,
```

Comment:
Now, `comment_payout_beneficiaries::validate()` and `comment_options_extension_visitor::operator()` returns different errors on same cases.
Errors should be the same.